sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/RedistributeTest.java

```diff
@@ -0,0 +1,396 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import static org.apache.beam.sdk.TestUtils.KvMatcher.isKv;
+import static org.apache.beam.sdk.util.construction.PTransformTranslation.REDISTRIBUTE_ARBITRARILY_URN;
+import static org.apache.beam.sdk.util.construction.PTransformTranslation.REDISTRIBUTE_BY_KEY_URN;
+import static org.apache.beam.sdk.values.TypeDescriptors.integers;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.collection.IsIterableContainingInAnyOrder.containsInAnyOrder;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertEquals;
+
+import java.io.Serializable;
+import java.util.List;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.coders.VarIntCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.testing.TestStream;
+import org.apache.beam.sdk.testing.UsesTestStream;
+import org.apache.beam.sdk.testing.ValidatesRunner;
+import org.apache.beam.sdk.transforms.Redistribute.AssignShardFn;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindow;
+import org.apache.beam.sdk.transforms.windowing.PaneInfo;
+import org.apache.beam.sdk.transforms.windowing.Sessions;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.construction.PTransformTranslation;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TimestampedValue;
+import org.apache.beam.sdk.values.TypeDescriptors;
+import org.apache.beam.sdk.values.ValueInSingleWindow;
+import org.apache.beam.sdk.values.WindowingStrategy;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableList;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Lists;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link Redistribute}. */
+@RunWith(JUnit4.class)
+public class RedistributeTest implements Serializable {
```

Comment:
Ah yes, I translated all tests of `Reshuffle.of()` to `Redistribute.arbitrariliy()`. FWIW I think letting the runner control the redistribute is definitely the best technical direction so that is the primary use case. But I change one back so we have some explicit coverage. I could add more.

The invariant for both transforms is simply that the output equals the input.

Reply:
I would suppose, that `Redistribute.byKey()` has additional semantics, or what would be the difference from `.arbitrarily()`? From my understanding the post-redistribute PCollection has data with same key co-located on the same partitions (workers). Is that not the case?

Reply:
`Redistribute.byKey()` lets the user include keys as a hint for the runner. In both cases, the transform should have output equal to input.

Reply:
If there are no semantic guarantees regarding how the key will be used then it's probably fine. Personally I'd prefer to implement `.arbitrarily()` using `keyBy()` with key as hash of value, but the fact that keys are co-located in partitions cannot be (currently) leveraged, so it probably does not matter.

FYI, I (curiously completely independently) found a bug in Flink's current batch reshuffle: https://github.com/apache/beam/pull/30622.

Reply:
Yea I agree with what you are saying. Some runners may have even better ways to redistribute arbitrarily (such as using dynamic runtime information to rebalance as it executes). But for Flink now I totally agree with your change, and I will try to mimic or leverage it.

Reply:
I got more information about the original change from `.rebalance()` to the hint. The problem was actually a bug in Flink: https://lists.apache.org/thread/tsbyzlk8p7m59qbgtgfm4rl0cx1rnz1j

I will try to get more info if this has been resolved and if we can use simple `rebalance()`.

Reply:
After offline discussion with @dmvk I think we can try to return back to the hint, but use the `_HASH` variant. According to the discussion this should be respected and have the required semantics (i.e. `Reshuffle` + `Redistribute.arbitrarily()`). I will try it next week.

Reply:
For this PR I hit some snags in the portable runner in every case so I just want to revert to copying reshuffle and we can fix it later. I updated it to match your new reshuffle implementation, but only the one that you modified. The rest are just what they already were.

Reply:
Thanks! I still wonder if we can cover the difference between `Redistribute.byKey()` and `.arbitrarily()` in tests? Or is this completely impossible? If it is impossible, why would one use the `.byKey()` variant?

Reply:
Both transforms are a hint. They are just different hints. We could run the equivalent tests for both, to make sure the runner doesn't mess one up when they directly implement. That is a good idea.